# Exploring eBay Car Sales Data
The aim of this project is to clean the data and analyze the included used car list

In [1]:
import numpy as np
import pandas as pd

Read the csv file into a dataframe

In [2]:
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

Print first and last 5 rows of the dataset

In [3]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.tail()

dateCrawled                                               name  \
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

      brand notRepairedDamage          dateCreated  nrOfPictures  postalCode  \
49995  audi              nein  2016-03-27 00:00:00             0       82131   
49996  opel              nein  2016-03-28 00:00:00             0       44807   
49997  fiat              nein  2016-04-02 00:00:00             0       73430   
49998  audi              nein  2016-03-08 00:00:00             0       35683   
49999  opel              nein  2016-03-13 00:00:00             0       45897   

                  lastSeen  
49995  2016-04-01 13:47:40  
49996  2016-04-02 14:18:02  
49997  2016-04-04 11:47:27  
49998  2016-04-05 16:45:07  
49999  2016-04-06 21:18:48

Get more information about datasets

In [5]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

- The dataset has 20 columns, most of which ares strings. The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores
- There are some columns that lack of values. So, we have to handle these columns latter:
    - vehicleType
    - gearbox
    - model
    - fuelType
    - notRepairedDamage  

## Cleaning Column Name
Take a look the columns names

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

Let rename some columnms first

In [7]:
new_name = {'yearOfRegistration':'registration_year',
           'monthOfRegistration': 'registration_month',
           'notRepairedDamage': 'unrepaired_damage',
           'dateCreated': 'ad_created'}
autos.rename(new_name, axis = 1, inplace = True)
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuelType', 'brand',
       'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

Convert the rest of the column names from camelcase to snakecase

In [8]:
def change_case(str):
     
    return ''.join(['_'+i.lower() if i.isupper()
               else i for i in str]).lstrip('_')
autos.columns = map(change_case, autos.columns)
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_p_s   model   odometer  registration_month fuel_type  \
0    manuell        158  andere  150,000km                   3       lpg   
1  automatik        286     7er  150,000km                   6    benzin   
2    manuell        102    golf   70,000km                   7    benzin   
3  automatik         71  fortwo   70,000km                   6    benzin   
4    manuell          0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

- We create a function to convert a string into snakecase format. You can easyly search more ways [in here](https://www.geeksforgeeks.org/python-program-to-convert-camel-case-string-to-snake-case/)  
- Use the map() functions to apply the convert function to each elements of the columns array  
- One last thing see here is the power_p_s column shoud be power_ps

In [9]:
autos.rename({'power_p_s': 'power_ps'}, axis = 1, inplace = True)
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

## Initial Exploration and Cleaning
Let's take a closer look our dataframe:
- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
- Examples of numeric data stored as text which can be cleaned and converted.

In [10]:
autos.describe()

registration_year      power_ps  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

There are some notes after looking at descriptive statistics for all columns:
- registration_year and power_ps should need more investigation since the max value of them is too illogically high
- nr_of_pictures mostly have one value so we can just remove it
- price and odometer columns stored as text that needs to be converted to numeric data

In [11]:
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].astype(int)
autos['price'].value_counts()


0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64

In [12]:
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].str.replace(',','')
autos['odometer'] = autos['odometer'].astype(int)
autos['odometer'].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer, dtype: int64

In [13]:
autos.rename({'odometer': 'odometer_km'}, axis = 1, inplace = True)
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [14]:
autos.describe()

price  registration_year      power_ps    odometer_km  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        50000.000000         50000.0  50000.000000  
mean             5.723360             0.0  50813.627300  
std              3.711984             0.0  25779.747957  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30451.000000  
50%              6.000000             0.0  49577.000000  
75%              9.000000             0.0  71540.000000  
max             12.000000             0.0  99998.000000

- We removed any non-numeric characters from price and odometer columns ('km', '$', ',')
- Next, we converted two columns to int type
- Finally, we renamed the odometer to odemeter_km and check the results by describe() function

## Exploring the Odometer and Price Columns

### odometer_km column

In [15]:
autos['odometer_km'].shape

(50000,)

In [16]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [17]:
autos['odometer_km'].value_counts().sort_index(ascending = True)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

The max value is 150000 that is consistent with current reality. So we do not need remove any row

### price column

In [18]:
autos['price'].shape

(50000,)

In [19]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [20]:
autos['price'].value_counts().sort_index(ascending = True)

0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price, Length: 2357, dtype: int64

- The max price of a car here is not reasonable. The [most expensive car in the world](https://heycar.co.uk/guides/most-expensive-cars-in-the-world) was just 70 millon dollar. So the upper bond is 70000000
- There are some price that too low, such as 0, 1, 2, 3, 5 dolar. So we need to look closer into these values. [The cheapest car](https://vocal.media/wheel/10-cheapest-cars-in-the-world) was just 3700 dollar. So let's try this value as lower bound

In [21]:
autos_0_price = autos[autos['price'] < 3700]
autos_0_price.shape

(28255, 20)

- 28255/50000 values is in this list. It's too much and will affect the analysis results if we remove them all
- In fact, we'll have to dig deeper into this data as to why the resale prices are so low. But in this problem I will take the lower bound as 500 dollar

In [22]:
autos_price = autos[autos['price'].between(500, 70000000)]

In [23]:
autos_price.shape

(45110, 20)

The new dataframe has 45110 rows. So 4890 outliers were removed

## Explorint the date columns

In [24]:
autos_price.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45110 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        45110 non-null  object
 1   name                45110 non-null  object
 2   seller              45110 non-null  object
 3   offer_type          45110 non-null  object
 4   price               45110 non-null  int64 
 5   abtest              45110 non-null  object
 6   vehicle_type        41432 non-null  object
 7   registration_year   45110 non-null  int64 
 8   gearbox             43292 non-null  object
 9   power_ps            45110 non-null  int64 
 10  model               43040 non-null  object
 11  odometer_km         45110 non-null  int64 
 12  registration_month  45110 non-null  int64 
 13  fuel_type           41906 non-null  object
 14  brand               45110 non-null  object
 15  unrepaired_damage   37588 non-null  object
 16  ad_created          45

- date_crawled, last_seen, and ad_created columns are all identified as string values by pandas, so we need to convert the data into a numerical
- registration_year and registration_month are represented as numeric values

In [25]:
autos_price[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

 The first 10 characters represent the day (e.g. 2016-03-12). we will calculate the distribution of values in the date_crawled, ad_created, and last_seen columns (all string columns) as percentages

### date_crawled column

In [26]:
autos_price['date_crawled'].str[:10]
autos_price['date_crawled'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05 14:06:30    0.000022
2016-03-05 14:06:40    0.000022
2016-03-05 14:07:08    0.000022
2016-03-05 14:07:21    0.000022
2016-03-05 14:07:26    0.000022
                         ...   
2016-04-07 14:30:09    0.000022
2016-04-07 14:30:26    0.000022
2016-04-07 14:36:44    0.000022
2016-04-07 14:36:55    0.000022
2016-04-07 14:36:56    0.000022
Name: date_crawled, Length: 43672, dtype: float64

### ad_created column

In [27]:
autos_price['ad_created'].str[:10]
autos_price['ad_created'].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11 00:00:00    0.000022
2015-08-10 00:00:00    0.000022
2015-09-09 00:00:00    0.000022
2015-11-10 00:00:00    0.000022
2015-12-05 00:00:00    0.000022
                         ...   
2016-04-03 00:00:00    0.039038
2016-04-04 00:00:00    0.037021
2016-04-05 00:00:00    0.011904
2016-04-06 00:00:00    0.003259
2016-04-07 00:00:00    0.001197
Name: ad_created, Length: 76, dtype: float64

### last_seen column

In [28]:
autos_price['last_seen'].str[:10]
autos_price['last_seen'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05 14:45:46    0.000022
2016-03-05 14:46:02    0.000022
2016-03-05 14:49:34    0.000022
2016-03-05 15:16:11    0.000022
2016-03-05 15:16:47    0.000022
                         ...   
2016-04-07 14:58:44    0.000067
2016-04-07 14:58:45    0.000022
2016-04-07 14:58:46    0.000022
2016-04-07 14:58:48    0.000067
2016-04-07 14:58:50    0.000067
Name: last_seen, Length: 35969, dtype: float64

### distribution of registration_year column

In [29]:
autos_price['registration_year'].describe()

count    45110.000000
mean      2005.061605
std         89.639687
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The max registration year can not be 9999. And the min year can not be 1000 as well. So we have to clean this column

## Dealing with Incorrect Registration Year Data

In [30]:
autos_final = autos_price[autos_price['registration_year'].between(1997, 2022)]

In [31]:
autos_final.shape

(40227, 20)

We choose the range of registration year from 1997 to 2022 aim at removing the illogical values. About 5000 value were removed from dataframe

In [32]:
autos_final['registration_year'].value_counts(normalize=True).sort_index()

1997    0.037587
1998    0.050364
1999    0.063962
2000    0.067492
2001    0.062669
2002    0.060134
2003    0.066050
2004    0.066696
2005    0.071196
2006    0.066299
2007    0.056430
2008    0.054839
2009    0.051682
2010    0.039426
2011    0.040197
2012    0.032491
2013    0.019813
2014    0.016233
2015    0.009049
2016    0.023616
2017    0.032242
2018    0.011510
2019    0.000025
Name: registration_year, dtype: float64

- The year has the highest number of car registrations is 2005
- The year has the smallest number of registrations is 2019
- The registrations increased sharply in the period from 1997 to 2009 and then gradually decreased

## Exploring Price by Brand

In [33]:
brands = autos_final['brand'].value_counts().sort_values(ascending = False)

In [34]:
unique_top_20_brand = brands[0:20].index

We selected top 20 brand from the dataset to aggregate the mean price of each brand by the following steps:
- Identify the unique values we want to aggregate by
- Create an empty dictionary to store our aggregate data
- Loop over the unique values, and for each: 
    - Subset the dataframe by the unique values 
    - Calculate the mean for each brand
    - Assign the val/mean to the dict as k/v

In [35]:
price_by_brand = {}
for b in unique_top_20_brand:
    df_temp = autos_final[autos_final['brand'] == b]
    price_by_brand[b] = df_temp['price'].mean()

In [36]:
price_by_brand

{'volkswagen': 6245.889592980746,
 'bmw': 9554.526374360115,
 'opel': 6475.076016357951,
 'mercedes_benz': 9120.766547861507,
 'audi': 10314.70497662909,
 'ford': 4426.956633590592,
 'renault': 2760.463326446281,
 'peugeot': 3388.5574940523393,
 'fiat': 15552.665330661323,
 'seat': 4824.809348093481,
 'skoda': 6558.925630810093,
 'smart': 3558.5532846715328,
 'mazda': 4521.142180094786,
 'citroen': 48187.59157212318,
 'nissan': 5325.778861788618,
 'toyota': 5516.545955882353,
 'hyundai': 5748.97995545657,
 'mini': 10689.833333333334,
 'volvo': 41869.99117647059,
 'kia': 6172.5}

In top 20 brand, citroen is the brand have the highest price, kia is the one have the lowest price

## Storing Aggregate Data in a DataFrame

In [37]:
price_by_brand_series = pd.Series(price_by_brand)
df = pd.DataFrame(price_by_brand_series, columns=['mean_price'])

In [38]:
df.head()

mean_price
volkswagen      6245.889593
bmw             9554.526374
opel            6475.076016
mercedes_benz   9120.766548
audi           10314.704977

We turn dictrionry into dataframe in 2 steps:
- Use pd.Series to convert dictionary into Series object
- Use pd.DataFrame to convert Series a to DataFrame